File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:

        res = ""
        maxlen = 0

        for i in range(len(s)):

            # odd
            l, r = i, i

            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r-l+1) > maxlen:
                    maxlen = r-l+1
                    res = s[l:r+1]
                l -= 1
                r += 1
        
            # even
            l, r = i, i+1

            while l >= 0 and r < len(s) and s[l] == s[r]:
                if (r-l+1) > maxlen:
                    maxlen = r-l+1
                    res = s[l:r+1]
                l -= 1
                r += 1

        return res



# memoization
# recursive - re-use
# controlled brute force
# sub problem

# single ch
# odd : left - center ch - right

# even : left == right

# 'aaa'
#  'aa'
# 'ac'
# 'aba'

File: 0005-longest-palindromic-substring/0005-longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        string res="";
        int longestlen = 0;
        int n=s.size();

        for(int i=0; i<n; i++) {

            // check odd size
            int l=i;
            int r=i;
            while(l>=0 && r<n && s[l]==s[r]) {
                if(longestlen < (r-l+1)) {
                    longestlen = r-l+1;
                    res = s.substr(l, r-l+1);
                }
                l--;
                r++;
            }
            l=i;
            r=i+1;
            while(l>=0 && r<n && s[l]==s[r]) {
                if(longestlen < (r-l+1)) {
                    longestlen = r-l+1;
                    res = s.substr(l, r-l+1);
                }
                l--;
                r++;
            }                        
        }
        return res;
    }
};

/*
sliding window
 babad
 l
  r
 
 cbbd
 l
 r
*/